test/recipes/15-test_dhparam.t

```diff
@@ -0,0 +1,97 @@
+
+./util/opensslwrap.sh genpkey -genparam -algorithm DH -pkeyopt pbits:3072 -pkeyopt qbits:160 -pkeyopt type:fips186_2 -pkeyopt dh_paramgen_type:0 -out $TESTDIR/p3072_q160_t1862_pgen0.pem
+./util/opensslwrap.sh genpkey -genparam -algorithm DH -pkeyopt pbits:3072 -pkeyopt qbits:224 -pkeyopt type:fips186_2 -pkeyopt dh_paramgen_type:0 -out $TESTDIR/p3072_q224_t1862_pgen0.pem
+./util/opensslwrap.sh genpkey -genparam -algorithm DH -pkeyopt pbits:3072 -pkeyopt qbits:256 -pkeyopt type:fips186_2 -pkeyopt dh_paramgen_type:0 -out $TESTDIR/p3072_q256_t1862_pgen0.pem
+
+
+./util/opensslwrap.sh genpkey -genparam -algorithm DH -pkeyopt pbits:1024 -pkeyopt qbits:160 -pkeyopt type:fips186_2 -pkeyopt dh_paramgen_type:1 -out $TESTDIR/p1024_q160_t1862_pgen1.pem
+./util/opensslwrap.sh genpkey -genparam -algorithm DH -pkeyopt pbits:1024 -pkeyopt qbits:224 -pkeyopt type:fips186_2 -pkeyopt dh_paramgen_type:1 -out $TESTDIR/p1024_q224_t1862_pgen1.pem
+./util/opensslwrap.sh genpkey -genparam -algorithm DH -pkeyopt pbits:1024 -pkeyopt qbits:256 -pkeyopt type:fips186_2 -pkeyopt dh_paramgen_type:1 -out $TESTDIR/p1024_q256_t1862_pgen1.pem
+
+./util/opensslwrap.sh genpkey -genparam -algorithm DH -pkeyopt pbits:1024 -pkeyopt qbits:160 -pkeyopt type:fips186_4 -pkeyopt dh_paramgen_type:1 -out $TESTDIR/p1024_q160_t1864_pgen1.pem
+
+./util/opensslwrap.sh genpkey -genparam -algorithm DH -pkeyopt pbits:2048 -pkeyopt qbits:160 -pkeyopt type:fips186_2 -pkeyopt dh_paramgen_type:1 -out $TESTDIR/p2048_q160_t1862_pgen1.pem
+./util/opensslwrap.sh genpkey -genparam -algorithm DH -pkeyopt pbits:2048 -pkeyopt qbits:224 -pkeyopt type:fips186_2 -pkeyopt dh_paramgen_type:1 -out $TESTDIR/p2048_q224_t1862_pgen1.pem
+./util/opensslwrap.sh genpkey -genparam -algorithm DH -pkeyopt pbits:2048 -pkeyopt qbits:256 -pkeyopt type:fips186_2 -pkeyopt dh_paramgen_type:1 -out $TESTDIR/p2048_q256_t1862_pgen1.pem
+
+./util/opensslwrap.sh genpkey -genparam -algorithm DH -pkeyopt pbits:2048 -pkeyopt qbits:224 -pkeyopt type:fips186_4 -pkeyopt dh_paramgen_type:1 -out $TESTDIR/p2048_q224_t1864_pgen1.pem
+./util/opensslwrap.sh genpkey -genparam -algorithm DH -pkeyopt pbits:2048 -pkeyopt qbits:256 -pkeyopt type:fips186_4 -pkeyopt dh_paramgen_type:1 -out $TESTDIR/p2048_q256_t1864_pgen1.pem
+
+./util/opensslwrap.sh genpkey -genparam -algorithm DH -pkeyopt pbits:3072 -pkeyopt qbits:160 -pkeyopt type:fips186_2 -pkeyopt dh_paramgen_type:1 -out $TESTDIR/p3072_q160_t1862_pgen1.pem
+./util/opensslwrap.sh genpkey -genparam -algorithm DH -pkeyopt pbits:3072 -pkeyopt qbits:224 -pkeyopt type:fips186_2 -pkeyopt dh_paramgen_type:1 -out $TESTDIR/p3072_q224_t1862_pgen1.pem
+./util/opensslwrap.sh genpkey -genparam -algorithm DH -pkeyopt pbits:3072 -pkeyopt qbits:256 -pkeyopt type:fips186_2 -pkeyopt dh_paramgen_type:1 -out $TESTDIR/p3072_q256_t1862_pgen1.pem
```

Comment:
Should we have some tests with `-algorithm DHX`?

Reply:
I think yes. Are the parameters documented/same?

Reply:
Actually its a bit confusing. DH is PKCS#3 style DH, while DHX is X9.42 style DH. But then its not clear what dh_paramgen_type does, since that is documented as meaning 0 == PKCS3 and 1 == X9.42. What happens if you do "-algorithm -DHX" but with "dh_paramgen_type:0", vs "-algorithm DH" with "dh_paramgen_type:1" what kind of params do you get (note the PEM headers are different for DH vs DHX)

Reply:
I suppose the documentation should be updated somehow, but I don't know what should be written here both for DH(X) and DSA parameters. Should I raise another issue for it?

Reply:
Its also unclear from this test what "pbits" and "qbits" do. I mean I intuitively understand what they do, but qbits is only documented for DSA not DH on the genpkey page, and pbits isn't documented at all except it appears in one of the genpkey examples (for DSA). So there are definitely documentation problems here - and a new issue for it would be a good idea.

Reply:
IMO the fact that dh_paramgen_type:1 is ignored is clearly a bug and one that should be fixed before beta1.

Reply:
And of course what does qbits mean for PKCS3 DH (which doesn't have a q parameter)?

Reply:
I hope it is just ignored...